### DISCUSSION TITLE
Chat workflow

### DISCUSSION THREAD

**Discussion by niba** [olimorris/codecompanion.nvim#82] [General]
I want to discuss how you use your code companion chats in neovim. I feel that experience could be better, but maybe I'm doing something wrong or it's impossible in neovim

Scenario:

I open one of my prompts to help me write a piece of code, and I have two windows

```
-----------------------
|  Code  |    Chat #1  |
-----------------------
```
At some point I want to open another chat window to do an alternative task. In my case, this is often generating a commit diff.

```
--------------------------------------------
|  Code  |    Chat #1  |  Chat Commit Diff |
--------------------------------------------
```
Now I have the following issues.

- `Chat toggle` doesn't work anymore for chat `#1`.
- My space is too small to have 3 windows open.

Let's assume that I generate commit diff, perform commit and close my `Chat Commit Diff`

Now i'm again in this state

```
-----------------------
|  Code  |    Chat #1  |
-----------------------
```

but my `Chat Toggle` doesn't work anymore.  I don't know how you can hide chat buffer without destroying it. Closing window just destroys the chat.

I have been using CodeCompanion for some time and I think that my ideal model looks like this

- Only one open chat (always)
- Opening a new chat hides the old one. You have keymaps to jump to old / new chats.
- Closing the window doesn't destroy the chat buffer. You have dedicated action if you want to remove chat. 
- Telescope with all chats (we almost have it) where you have preview of chat content and can perform actions like destroy and save chat. I don't use other "actions" from code companion popup

**Comment by olimorris** [comment]
Thanks for your feedback. This is exactly what I’ve been looking for. I am not sure how others are using CodeCompanion and if they’re using it anywhere near the same as me 😂.

Some great points about having multiple chats open. Rare that I have that scenario but totally see how you could end up with it. Love the idea of being able to move between chats with a key map. You sort of can hide a chat buffer with the toggle command but appreciate there could easily be a hide keymap. A lot of the functionality exists in the chat buffer itself.

Do you have the chat open in a split? What keymappings have you setup for yourself and what do you use in the chat buffer?

**Comment by niba** [comment]
The problem with the `toggle` command is that it only works for the newest chat. If I have `Chat #1` and I open `Chat #2` then the toggle hides `Chat #2`. I can't use toggle anymore to hide `Chat #1`.  It doesn't matter which window is focused. When I destroy `Chat #2` then toggle opens a new chat. It still doesn't have any effect on `Chat #1` and in this situation i can only destroy it to hide it 😄 

I think that most people don't want to have more than one chat window open. That's why I would much prefer to have always one chat (it could be an option). This design also fixes the toggle functionality because, if you have only one chat window visible then toggle always works correctly.

Another thing is that it's currently too easy to accidentally destroy a chat. I'm used to native bindings to manage my windows so I often destroy chat window even If I don't want to.

I always open chats in a vertical split. I don't use inline code generation, agents and workflows.

I use chats as my documentation or source of knowledge. I don't like when AI writes my code. I just prefer to use is as a helper. That's why I often don't want to clear or destroy my chats because I might need this knowledge in the near future.

Scenario
- At work, we are adopting a new library `xstate`. I use chat to help me understand some library concepts and model my data in a correct way
- Then I need to jump to a different part of my system to write communication between iframes and again I try to use chat to help me improve my typescript types or help with iframe API. At this point I don't want to mix my `xstate` conversation with the `iframe` conversation so I create a new chat.
- Later, I go back to modeling my `xstate` code and want to bring back the old `xstate` conversation.

Right now, achieving this kind of workflow is quite problematic. I basically just "save chat" every time but this is far from ideal. That's why I suggested this workflow with a single chat window and the ability to move through chat conversations using keybindings or a telescope. 

I use `<leader>a` as my section for AI.
All my prompts are in `<leader>ap` region.

```lua
    keys = {
      {
        "<leader>aa",
        "<cmd>CodeCompanionToggle<cr>",
        desc = "Toggle Chat",
        mode = { "n", "v" },
      },
      {
        "<leader>ao",
        "<cmd>CodeCompanionActions<cr>",
        desc = "Options",
        mode = { "n", "v" },
      },
      {
        "<leader>av",
        "<cmd>CodeCompanionAdd<cr>",
        desc = "Add visual selected",
        mode = { "v" },
      },
    },
    keymaps = {
      -- reset original bindings (conflicts with existing lazyvim bindings)
      ["gc"] = "",
      ["ga"] = "",
      ["gs"] = "",
      ["gt"] = "",
      ["<C-c>"] = "<esc>",

      ["<C-s>"] = "keymaps.save",
      ["<C-x>"] = "keymaps.close",
      ["q"] = "keymaps.stop",
      ["<leader>agc"] = "keymaps.clear", -- don't use (i just open new one)
      ["<leader>aga"] = "keymaps.codeblock",
      ["<leader>ags"] = "keymaps.save_chat",
      ["<leader>agt"] = "keymaps.add_agent", -- don't use
      ["]"] = "keymaps.next", -- don't use
      ["["] = "keymaps.previous", -- don't use
    },
    -- some of prompts, i have few more but the concept is the same
    ["Generate a Commit Message"] = {
      opts = {
        mapping = "<leader>apc",
      },
    },
    ["Explain LSP Diagnostics"] = {
      opts = {
        mapping = "<leader>apl",
      },
    },
    ["Code Advisor"] = {
      opts = {
        mapping = "<leader>apa",
      },
    },
    ["Senior Developer"] = {
      opts = {
        mapping = "<leader>aps",
      },
    },

```

**Reply by olimorris** [reply]
Great explanation and appreciate you sharing your workflow. In total agreement with this and will put it on the roadmap for this week.

As a heads up I changed the keymap config massively as of yesterday.

**Reply by niba** [reply]
Awesome to hear that. Thanks for your hard work and for creating a great AI plugin. It's the best one in the Neovim community.

**Reply by olimorris** [reply]
@niba let me know your thoughts on #88
